### PR TITLE
main: fix #15858 by setting dropPeerFn in copydb

### LINKED ISSUE #15858: Unable to perform copydb (panic: runtime error: memory)
#### System information

Geth version: `Geth/v1.7.3-stable-4bb3c89d/linux-amd64/go1.9`
OS & Version: Ubuntu 16.04.3 x64

I was following @karalabe 's instructions in #15797 and I got to block number 2433095 when geth crashed:

Command I ran `geth --datadir=/gethswap copydb --cache=512 /geth/.ethereum/geth/chaindata`

````
INFO [01-11|15:34:00] Imported new block receipts              count=283  elapsed=128.139ms bytes=817740   number=2433095 hash=65693e…3957ec ignored=0

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xa45f88]

goroutine 58 [running]:
github.com/ethereum/go-ethereum/eth/downloader.(*Downloader).fetchHeaders(0xc420216680, 0xc42025ce10, 0x258001, 0x0, 0x0)
        /build/ethereum-ZSPmrv/ethereum-1.7.3+build11486+xenial/build/_workspace/src/github.com/ethereum/go-ethereum/eth/downloader/downloader.go:859 +0xf18
github.com/ethereum/go-ethereum/eth/downloader.(*Downloader).syncWithPeer.func3(0x8, 0xffd8b0)
        /build/ethereum-ZSPmrv/ethereum-1.7.3+build11486+xenial/build/_workspace/src/github.com/ethereum/go-ethereum/eth/downloader/downloader.go:474 +0x40
github.com/ethereum/go-ethereum/eth/downloader.(*Downloader).spawnSync.func1(0xc42c1a5ee0, 0xc420062c60, 0xc42851f3a0)
        /build/ethereum-ZSPmrv/ethereum-1.7.3+build11486+xenial/build/_workspace/src/github.com/ethereum/go-ethereum/eth/downloader/downloader.go:500 +0x51
created by github.com/ethereum/go-ethereum/eth/downloader.(*Downloader).spawnSync
        /build/ethereum-ZSPmrv/ethereum-1.7.3+build11486+xenial/build/_workspace/src/github.com/ethereum/go-ethereum/eth/downloader/downloader.go:500 +0xc7

root@GETH01:/# free -h
              total        used        free      shared  buff/cache   available
Mem:           7.8G         64M        2.3G        516K        5.4G        7.4G
Swap:          4.0G         30M        4.0G

root@GETH01:/# df -h
Filesystem      Size  Used Avail Use% Mounted on
udev            3.9G     0  3.9G   0% /dev
tmpfs           797M  8.8M  788M   2% /run
/dev/sda1       193G  174G  9.8G  95% /
tmpfs           3.9G     0  3.9G   0% /dev/shm
tmpfs           5.0M     0  5.0M   0% /run/lock
tmpfs           3.9G     0  3.9G   0% /sys/fs/cgroup
/dev/sdb1        99G  4.6G   89G   5% /gethswap
tmpfs           797M     0  797M   0% /run/user/1000
````

### PR DESCRIPTION
**Pull request by holiman**
Fixes https://github.com/ethereum/go-ethereum/issues/15858 by setting the peer-drop-function to non-nil.

### CODE REVIEW

**Comment by holiman** [comment]
Test:
```
INFO [01-30|09:02:23] Imported new block receipts              count=4   elapsed=123.965µs bytes=1096 number=6   hash=1f1aed…6b326e ignored=0
WARN [01-30|09:02:23] Downloader tried to fake-peer; ignoring. peer=local
INFO [01-30|09:02:23] Imported new block headers               count=192 elapsed=29.385ms  number=384 hash=d3d5d5…c79cf3 ignored=0

```

File: cmd/geth/chaincmd.go [hunk 1]
```diff
@@ -297,7 +297,10 @@ func copyDb(ctx *cli.Context) error {
 	chain, chainDb := utils.MakeChain(ctx, stack)
 
 	syncmode := *utils.GlobalTextMarshaler(ctx, utils.SyncModeFlag.Name).(*downloader.SyncMode)
-	dl := downloader.New(syncmode, chainDb, new(event.TypeMux), chain, nil, nil)
+
+	dl := downloader.New(syncmode, chainDb, new(event.TypeMux), chain, nil, func(id string) {
+		log.Warn("Downloader tried to fake-peer; ignoring.", "peer", id)
```
Comment (karalabe) [inline cmd/geth/chaincmd.go]: "fake" -> "drop"?

Comment (holiman) [reply cmd/geth/chaincmd.go]: ...doh

**Comment by karalabe** [comment]
All in all I'm a bit unsure about this PR:

 * If we're syncing with ourselves, why do we want to drop? That seems to point to a bug in the downloader and which should rather be fixed.
 * Also my gut feeling is that this PR would break some internal downloader logic which assumes that the peer will be gone and can be cleaned up after it vs. this code just ignoring it. I'm afraid of some unforeseen side effects.

**Comment by holiman** [comment]
> If we're syncing with ourselves, why do we want to drop? 

In my case, it occurred when I was doing a measurement using @fjl's memsize tool, which stops everything and scans memory for a minute or so. I'm assuming that when it starts again, the downloader suddenly finds that the time has progressed, and that the local peer obviously hasn't responded in time. 

For the other cases, I'm assuming disk IO problems. 

>  I'm afraid of some unforeseen side effects.

Well, worse than `panic` ?

**Comment by karalabe** [comment]
> For the other cases, I'm assuming disk IO problems.

Fair point.

> Well, worse than panic ?

Ok, I can live with this PR if you can "prove" that a noop drop won't corrupt the downloader. I.e. please skim through the downloader and make sure there won't be something going wrong.

**Comment by holiman** [comment]
> Ok, I can live with this PR if you can "prove" that a noop drop won't corrupt the downloader. I.e. please skim through the downloader and make sure there won't be something going wrong.

I added a snippet so that the drop-case runs immediately after it receives a packet. It stops importing headers/receipts, and continues only downloading states... myeah it's not great...

**Comment by holiman** [comment]
So if I actually do drop it, we get this: 
```
INFO [01-30|10:21:27] Imported new block headers               count=1344 elapsed=161.532ms number=27648 hash=e3446d…019e62 ignored=0
WARN [01-30|10:21:27] Downloader wants to drop fake-peer       peer=local
WARN [01-30|10:21:27] Node data write error                    err="state node feab52…7a51be failed with all peers (0 tries, 0 peers)"
INFO [01-30|10:21:27] Imported new block headers               count=2048 elapsed=159.505ms number=29696 hash=1d7dd6…675b9f ignored=0
WARN [01-30|10:21:28] Rolled back headers                      count=2048 header=29696->27648 fast=26304->26304 block=0->0
WARN [01-30|10:21:28] Synchronisation failed, dropping peer    peer=local err="no peers available or all tried for download"
WARN [01-30|10:21:28] Downloader wants to drop fake-peer       peer=local
ERROR[01-30|10:21:28] Failed to unregister sync peer           peer=local err="peer is not registered"
no peers available or all tried for download
```

**Comment by holiman** [comment]
Ok, maybe this is the best solution:
```diff
diff --git a/cmd/geth/chaincmd.go b/cmd/geth/chaincmd.go
index e094484..918ce9e 100644
--- a/cmd/geth/chaincmd.go
+++ b/cmd/geth/chaincmd.go
@@ -298,9 +298,7 @@ func copyDb(ctx *cli.Context) error {
 
        syncmode := *utils.GlobalTextMarshaler(ctx, utils.SyncModeFlag.Name).(*downloader.SyncMode)
 
-       dl := downloader.New(syncmode, chainDb, new(event.TypeMux), chain, nil, func(id string) {
-               log.Warn("Downloader tried to drop fake-peer; ignoring.", "peer", id)
-       })
+       dl := downloader.New(syncmode, chainDb, new(event.TypeMux), chain, nil,nil)
 
        // Create a source peer to satisfy downloader requests from
        db, err := ethdb.NewLDBDatabase(ctx.Args().First(), ctx.GlobalInt(utils.CacheFlag.Name), 256)
[~/go/src/github.com/ethereum/go-ethereum]
#git diff eth/downloader/downloader.go
diff --git a/eth/downloader/downloader.go b/eth/downloader/downloader.go
index b338129..1c393d3 100644
--- a/eth/downloader/downloader.go
+++ b/eth/downloader/downloader.go
@@ -853,6 +853,11 @@ func (d *Downloader) fetchHeaders(p *peerConnection, from uint64) error {
                        getHeaders(from)
 
                case <-timeout.C:
+
+                       if d.dropPeer == nil{
+                               p.log.Warn("Downloader wants to drop peer, but peerdrop-function is not set")
+                               break
+                       }
                        // Header retrieval timed out, consider the peer bad and drop
                        p.log.Debug("Header request timed out", "elapsed", ttl)
                        headerTimeoutMeter.Mark(1)
```

**Comment by holiman** [comment]
Ok, updated the fix to check if function is `nil` instead, and if so just log a `warn` and ignore the timeout

File: eth/downloader/downloader.go [hunk 2]
```diff
@@ -853,6 +856,11 @@ func (d *Downloader) fetchHeaders(p *peerConnection, from uint64) error {
 			getHeaders(from)
 
 		case <-timeout.C:
+
```
Comment (karalabe) [inline eth/downloader/downloader.go]: Please drop this empty line here.

**Comment by holiman** [comment]
Addressed concerns, and squashed ☑️

**Pull request merged**